rfcs/0039-minaform-tool-design.md

```diff
@@ -0,0 +1,117 @@
+## Summary
+[summary]: #summary
+
+This rfc proposes a new tool dubbed "minaform" for reliably creating and managing QA nets and production deployments. Our existing setup is very hacky and requires knowledge of many tools to use effectively. The intention with minaform is to build a flexible enough tool that resolves major pain points with the existing setup, and standardizes a single tool for folks to learn/understand.
+
+## Motivation
+[motivation]: #motivation
+
+Why are we doing this? What use cases does it support? What is the expected outcome?
+
+The following issues with the existing system were reported by at least one person as "pain points" that consume time/debugging effort when deploying networks:
+
+1) Production Networks vs. QA Nets
+    - The process for deploying produciton testnets vs. qa nets has diverged in a few places and this means our day-to-day testing does not cover all of the services we need to deploy for public networks.
+2) Determining network health is difficult/unclear/inconsistent
+    - Some of this is addressed with more health checks in k8s and more metrics / better dashboards in grafana, but integrating the deployment tooling with all of those checks can help make it more foolproof.
+3) No "uninstall" command / cleanup afterwards
+    - Both the testnet terraform files and resources in the cluster end up hanging around more than neccessary. Needs additional tooling to clean up after itself.
+4) Generating Genesis Proofs is slow/resource intensive/hard to predict
+    - See proposal for how to pre-generate proofs in docker images
+5) Hard to determine the correct docker tag from a given git SHA
+    - Can be scripted to allow for a git tag or git SHA input, in addition to / in place of the docker tag.
+6) Keys and other private info for deployment are not committed/shared with other team members
+    - Addressed by https://github.com/MinaProtocol/coda-automation/issues/472
+7) Incompatible helm/terraform versions
+    - Can be addressed by 1) checking versioning before trying to use these tools and 2) bundling a canonical set of tools into a docker image
+8) Incompatible helm/terraform changes
+    - Addressed by moving the charts to the mina repo and explicitly releasing them. Terraform specifies an individual version of each chart to depend on and will fetch the released chart before deploys.
+9) Helm string processing is inconsistent/hard
+    - Now all charts are passing `helm lint` in the mina repo, and we can add `helm test` in the future as well.
+10) Passing flags to the daemon process directly is difficult to debug/handle/manage
+    -  Move to using daemon.json more and exposing all of these flags through the charts.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+The design revolves around a new bash script, minaform, that can be bundled in a container alongside its dependencies.
```

Comment:
Why `bash` over another language? If this tool is to have a robust CLI interface and be used in a monolithic fashion, `bash` seems like a difficult choice, as not everyone is familiar with it enough to write it effectively, and there's not easy way to generate a robust CLI interface with both documentation and consistent arg parsing logic. Other languages, such as `python`, `reasonml`, and `ocaml` all have libraries for generating the CLI interface from higher level representations.